### ISSUE TITLE
I've updated my benchmark for 3 active js Physics Engines.

### ISSUE THREAD

**Issue by finscn** [shakiba/planck.js#207]
I've updated my benchmark for 3 active js Physics Engines.
This benchmark is very very simple  and basic .
But I think it has already could tell us something .

https://finscn.github.io/phy-benchmark/

Bad news , planck.js is the slowest one

**Comment by zOadT** [comment]
Thank you very much for reporting!

I will try to find and remove deoptimizations as soon as I find some time

**Comment by 8Observer8** [comment]
Asus K53SV; 8 GB RAM, i3 2.2 GHertz (2 cores); Intel HD Graphics 3000; Nvidia Geforce GT 540M (1 GB); Windows 10

- box2d.ts : 25 FPS
- planck.js : 16 FPS
- matter.js : 42 FPS

**Comment by quinton-ashley** [comment]
matter.js isn't really comparable, even though it is faster the quality of its physics simulation is much worse.

**Comment by 8Observer8** [comment]
I found another benchmark here: https://lusito.github.io/box2d.ts/benchmark/

On my laptop ASUS K53SV:

**Browser:** Chrome 104
| Name           | avg ms/frame |           5th %ile |          95th %ile | Ratio |
| -------------- | ------------ | ------------------ | ------------------ | ----- |
| box2d.js       |         7.53 |  8.099999904632568 | 13.600000143051147 |  1.00 |
| box2d-web      |        14.19 |               15.5 | 12.599999904632568 |  1.89 |
| @box2d/core    |        20.95 | 21.299999952316284 | 16.799999952316284 |  2.78 |
| box2d-html5    |        27.44 | 28.200000047683716 | 26.399999856948853 |  3.65 |
| @flyover/box2d |        39.47 | 41.799999952316284 |                 38 |  5.24 |
| planck.js      |        56.85 |                 61 | 44.799999952316284 |  7.55 |

**Comment by quinton-ashley** [comment]
Stats on my M1 MacBook Air 2020

**Browser:** Chrome 104
| Name           | avg ms/frame |           5th %ile |          95th %ile | Ratio |
| -------------- | ------------ | ------------------ | ------------------ | ----- |
| box2d.js       |         5.43 |                  6 |  6.099999999627471 |  1.00 |
| box2d-web      |         5.87 |  5.599999999627471 | 6.1000000005587935 |  1.08 |
| @box2d/core    |         6.35 |  6.099999999627471 | 7.2000000001862645 |  1.17 |
| box2d-html5    |         7.07 | 7.3999999994412065 |                  7 |  1.30 |
| @flyover/box2d |        10.50 | 11.199999999254942 | 10.200000000186265 |  1.93 |
| planck.js      |        13.04 |  26.09999999962747 |               10.5 |  2.40 |

**Browser:** Firefox 104
| Name           | avg ms/frame | 5th %ile | 95th %ile | Ratio |
| -------------- | ------------ | -------- | --------- | ----- |
| box2d.js       |         5.15 |        6 |         5 |  1.00 |
| box2d-web      |         7.12 |        7 |         6 |  1.38 |
| @box2d/core    |        10.41 |       11 |         7 |  2.02 |
| box2d-html5    |        10.73 |       12 |         9 |  2.08 |
| planck.js      |        20.66 |       22 |        24 |  4.01 |
| @flyover/box2d |        23.38 |       31 |        23 |  4.54 |

**Comment by quinton-ashley** [comment]
In case anyone else stumbles across this thread, currently the best version of box2d with the most features and best performance is [liquidfun-play-2](https://github.com/Birch-san/liquidfun-play-2). It runs on web browsers via WASM which is why it's so fast.

**Comment by 8Observer8** [comment]
I see that `liquidfun-play-2` is using [box2d-wasm](https://www.npmjs.com/package/box2d-wasm?activeTab=readme). I'm using it now too because it has the b2Draw class, but it's not in planck.js. But box2d-wasm only has 32 downloads per week.

My simple example for beginners in Box2D like me: https://plnkr.co/edit/exqhktfXnvxwtBRq

![214072546-92d62972-4066-4025-aa6a-afa94434bf22](https://user-images.githubusercontent.com/3908473/214592601-ed66f81e-8c06-4d73-84e3-aa1f8f2129cc.gif)

**Comment by quinton-ashley** [comment]
@8Observer8 Nice! What's the reason you multiplied the time step (1/60) by 1000?

**Comment by 8Observer8** [comment]
Yes you are right) I just copied the step logic from here: https://github.com/Birch-san/box2d-wasm/blob/master/demo/modern/public/demo.js#L96

**Comment by 8Observer8** [comment]
I think it is because `const nowMs = window.performance.now()` returns milliseconds:

```
    (function animationLoop(prevMs) {
        const nowMs = window.performance.now()
        window.requestAnimationFrame(animationLoop.bind(null, nowMs));
        const deltaMs = nowMs - prevMs;
        step(deltaMs);
        draw();
    })(window.performance.now());
```

**Comment by quinton-ashley** [comment]
Oh wow how cool. So it's changing the time step depending on how long the previous frame took to render?

**Comment by 8Observer8** [comment]
>  it's changing the time step depending on how long the previous frame took to render

Unfortunately, I don't fully understand myself. What do you think about it?

**Comment by 8Observer8** [comment]
I think [@dimforge/rapier2d-compat](https://www.npmjs.com/package/@dimforge/rapier2d-compat) should be included in testing.

**Comment by quinton-ashley** [comment]
Yeah that's what it's doing but not for the reason I thought. 1/60 is the slowest speed the simulation can run at but if the user has a high frame rate monitor it could run the simulation faster with a lower time step.

Does Rapier use box2d? If so then yeah it'd be nice to see a comparison between Rapier WASM and liquidfun-play-2!

**Comment by 8Observer8** [comment]
> Does Rapier use box2d?

Rapier2D is another physics engine written in Rust: https://github.com/dimforge/rapier and ported to JS and TS: https://github.com/dimforge/rapier.js.

Rapier2D has collider debug rendering: https://github.com/dimforge/rapier.js/pull/119 But I don't know about its performance.

**Comment by 8Observer8** [comment]
> it'd be nice to see a comparison between Rapier WASM and liquidfun-play-2!
Yes, me too.

https://rapier.rs/docs/user_guides/javascript/getting_started_js/

> Rapier is actually a WebAssembly module

**Comment by quinton-ashley** [comment]
Could you add liquidfun-play-2 and rapier to the benchmark test you made?

**Comment by 8Observer8** [comment]
> Could you add liquidfun-play-2 and rapier to the benchmark test you made?

This is not my test. This was done by Lusito, the owner of this repository:: https://github.com/Lusito/box2d.ts

**Comment by quinton-ashley** [comment]
ah I see

**Comment by finscn** [comment]
Hi I've update this test page. the planck.js is still worse than box2d.ts .
the test code  at https://github.com/finscn/finscn.github.io/tree/master/phy-benchmark 
Is there something wrong in my test case ?

**Comment by 8Observer8** [comment]
On my laptop the best one is Matter.js - 60 FPS

box2d.ts: http://fatidol.com/phy-benchmark/box2d-ts-test.html
![image](https://github.com/piqnt/planck.js/assets/3908473/2c3f8521-2458-4778-9957-24ae6996132a)

Planck.js: http://fatidol.com/phy-benchmark/planck-test.html
![image](https://github.com/piqnt/planck.js/assets/3908473/13934698-41b1-40ad-8895-d2ed667f95bf)

Matter.js: http://fatidol.com/phy-benchmark/matter-test.html
![image](https://github.com/piqnt/planck.js/assets/3908473/3f396dd2-6a58-4e07-a395-88f58ceb9245)

**Comment by quinton-ashley** [comment]
@8Observer8 matter.js is fast but isn't the best though cause the quality of it's physics simulation isn't comparable to box2d.

**Comment by photonstorm** [comment]
They all run at a solid 60fps for me (iMac M1), but the Matter simulation has obvious issues with bodies penetrating each other. Maybe a more intensive test would be helpful?

**Comment by 8Observer8** [comment]
Thanks guys for the warning, otherwise I was already thinking about taking Matter.js as it is twice as fast as box2d.ts.

**Comment by quinton-ashley** [comment]
@photonstorm @8Observer8 @finscn You guys should check out p5play v3. I made tons of user friendly wrappers around planck.js.
https://p5play.org/learn/sprite.html

For v4 of p5play I plan to use box2dwasm behind the scenes.

**Comment by shakiba** [comment]
@finscn Thanks for updating us with the result!

**Comment by shakiba** [comment]
> But in my games there will be few objects, so in terms of performance they are the same for me.

@8Observer8 That's true, this is the case about most games. There are specific cases where performance matters, and because planck.js has readable JS code a developer can adjust the library for their use-case. 

I wonder what is the use-case for b2Draw in JS?

**Comment by shakiba** [comment]
Another thing to mention is that you can reduce physics frame-rate, and use transitions for rendering objects at higher rate, so it won't have any impact on user experience.

Box2D/Planck is designed to work at a fixed frame rate anyway, unlike requestAnimationFrame which runs at variable rate and is supposed to run as fast as possible.

For example you can setup the physics to run at fixed 30 or 20 steps per second, and run rendering at any higher rate.
